piet-direct2d/src/lib.rs

```diff
@@ -555,20 +559,358 @@ impl<'a> TextLayoutBuilder for D2DTextLayoutBuilder<'a> {
     type Out = D2DTextLayout;
 
     fn build(self) -> Result<Self::Out, Error> {
-        Ok(D2DTextLayout(
-            self.builder
-                .with_text(&self.text)
-                .with_font(&self.format)
-                .with_width(1e6) // TODO: probably want to support wrapping
-                .with_height(1e6)
-                .build()
-                .wrap()?,
-        ))
+        Ok(D2DTextLayout {
+            layout: self.builder
+                    .with_text(&self.text)
+                    .with_font(&self.format)
+                    .with_width(1e6) // TODO: probably want to support wrapping
+                    .with_height(1e6)
+                    .build()
+                    .wrap()?,
+            text: self.text,
+        })
     }
 }
 
 impl TextLayout for D2DTextLayout {
     fn width(&self) -> f64 {
-        self.0.get_metrics().width() as f64
+        self.layout.get_metrics().width() as f64
+    }
+
+    fn hit_test_point(&self, point: Point) -> HitTestPoint {
+        // lossy from f64 to f32, but shouldn't have too much impact
+        let htp = self.layout.hit_test_point(
+            point.x as f32,
+            point.y as f32,
+        );
+
+        let text_position = htp.metrics.text_position() as usize;
+
+        // Convert text position from utf-16 code units to
+        // utf-8 code units.
+        // Strategy: count up in utf16 and utf8 simultaneously, stop when
+        // utf-16 text position reached.
+        //
+        // TODO ask about text_position, it looks like windows returns last index;
+        let idx_8 = count_until_utf16(&self.text, text_position)
+            .unwrap_or(self.text.len() - 1);
+            //.expect("(logic error?) utf16 text position not found in string");
+
+        HitTestPoint {
+            metrics: HitTestMetrics {
+                text_position: idx_8,
+                is_text: htp.metrics.is_text(),
+            },
+            is_inside: htp.is_inside,
+            is_trailing_hit: htp.is_trailing_hit,
+        }
+    }
+
+    fn hit_test_text_position(&self, text_position: usize, trailing: bool) -> Option<HitTestTextPosition> {
+        // now convert the ut8 index to utf16
+        // TODO this can panic; should try next text_position
+        let idx_16 = count_utf16(&self.text[0..text_position]);
+
+        // panic or Result are also fine options for dealing with overflow. Using Option here
+        // because it's already present and convenient.
+        // TODO this should probably go before convertin to utf16, since that's relatively slow
+        let idx_16 = idx_16.try_into().ok()?;
+
+
+        // Check for idx out of bounds.
```

Comment:
Sure. Directwrite will return the line width if you try a textPosition that's beyond the end, so I'm doing an explicit check instead to get back None.

However, I now realize that this branch is taken only when `text_position = text.len()`, since that's the one time text position is both out of bounds and can be used as a boundary for the slice for counting utf16 code units. Any higher, and that slice will panic first.

Let me know if you have any thoughts on this, otherwise I'll figure out something consistent.

Reply:
Implemented: removed explicit check.